### PR TITLE
Update to new progress spec

### LINKED ISSUE #184: Update to current state of progress proposal
The proposal for reporting progress has changed quite significantly since it was added to `haskell-lsp`. There is now a branch for LSP 3.15 that contains the current state https://github.com/microsoft/language-server-protocol/blob/dbaeumer/3.15/specification.md so hopefully the current state is somewhat final and we can update haskell-lsp to use it.

### PR DESCRIPTION
**Pull request by cocreature**
fixes #184 

I have tested this in `ghcide` with the latest VSCode client libraries, see https://github.com/digital-asset/ghcide/pull/153 for the corresponding PR.

### CODE REVIEW

**Comment by cocreature** [comment]
The CI failure on `8.0.2` seems to be a stack bug see https://github.com/commercialhaskell/stack/issues/4951. It looks like it’s fixed in master already https://github.com/commercialhaskell/stack/pull/4955

**Comment by alanz** [comment]
cc @mpickering

File: haskell-lsp-types/src/Language/Haskell/LSP/Types/Message.hs [hunk 1]
```diff
@@ -83,7 +83,7 @@ data ClientMethod =
  | WorkspaceSymbol
  | WorkspaceExecuteCommand
  -- Progress
- | WindowProgressCancel
+ | WorkDoneProgressCancel
```
Comment (bubba) [inline haskell-lsp-types/src/Language/Haskell/LSP/Types/Message.hs]: Does `window/workDoneProgress/cancel` exist anymore in the 3.15 specification? Are we supposed to just use https://microsoft.github.io/language-server-protocol/specifications/specification-3-15/#cancelRequest now?

Comment (cocreature) [reply haskell-lsp-types/src/Language/Haskell/LSP/Types/Message.hs]: It still exists in https://github.com/microsoft/vscode-languageserver-node/blob/master/protocol/src/protocol.progress.proposed.ts#L127 so I assume that the omission from the spec was a mistake (given that other things like `workDoneProgress/create` are also not documented properly but I don’t know tbh.

File: haskell-lsp-types/src/Language/Haskell/LSP/Types/FoldingRange.hs [hunk 2]
```diff
@@ -7,12 +7,14 @@ import           Data.Aeson.TH
...
 import           Language.Haskell.LSP.Types.Constants
 import           Language.Haskell.LSP.Types.List
+import           Language.Haskell.LSP.Types.Progress
 import           Language.Haskell.LSP.Types.TextDocument
 import           Language.Haskell.LSP.Types.Message
 
 data FoldingRangeParams =
   FoldingRangeParams
   { _textDocument :: TextDocumentIdentifier -- ^ The text document.
+  , _workDoneToken :: Maybe ProgressToken -- ^ An optional token that a server can use to report work done progress.
```
[the remark this answers is not in the document]
Comment (cocreature) [reply haskell-lsp-types/src/Language/Haskell/LSP/Types/FoldingRange.hs]: Yeah that’s a good point, `TextDocumentPositionParams` covers a bunch of them but I haven’t tried to be exhaustive so far.

[the remark this answers is not in the document]
Comment (alanz) [reply haskell-lsp-types/src/Language/Haskell/LSP/Types/FoldingRange.hs]: Does it need an issue?

[the remark this answers is not in the document]
Comment (bubba) [reply haskell-lsp-types/src/Language/Haskell/LSP/Types/FoldingRange.hs]: Created #187

**Comment by bubba** [comment]
Thanks for adding this, this is great! I'm happy to merge this if everyone else is

**Comment by alanz** [comment]
Merging.

@bubba, feel free to make a release (together with lsp-test)

**Comment by bubba** [comment]
@alanz I think we should try and finish off #187 before making a release

**Comment by cocreature** [comment]
I’ll try to take a look at #187 next week if nobody beats me to it.

**Comment by bubba** [comment]
@cocreature I've just started a [branch](https://github.com/alanz/haskell-lsp/tree/more-WorkDoneProgressParams), but started running into some annoying problems with parameters that are just an extension of `TextDocumentPositionParams` and `WorkDoneProgressParams`, like [`HoverParams`](https://microsoft.github.io/language-server-protocol/specifications/specification-3-15/#textDocument_hover). 

There's a whole load of params that just emend the two, and at the moment we're just using `TextDocumentPositionParams` to represent them. I don't want to have to create a new data type for every single one and copy over the fields. I'm thinking we could create a `TextDocumentPositionAndWorkDoneProgressParams`, its a bit hairy though. Maybe we could use some template haskell for this?

**Comment by cocreature** [comment]
@bubba In this PR I’ve added the `workDoneToken` to `TextDocumentPositionParams`. Does that not work? I thought I looked through them and all of them now have a token but I might have missed something.

**Comment by bubba** [comment]
@cocreature ah I didn't notice that actually! That takes care of things then 👍

**Comment by bubba** [comment]
Feel free to fire away at that branch by the way!

**Pull request merged**